### ISSUE TITLE
Implement engine damage when flying at high RPM for long time

### ISSUE THREAD

**Issue by gilbertohasnofb** [c172p-team/c172p-detailed#124]
Implement engine damage when flying at a RPM higher than the red mark (2700 RPM) for long time.

**Comment by gilbertohasnofb** [comment]
@onox do you want to work on this? I had proposed to help with a dialog, similar to what I did for the oil and water contamination. What do you think?

**Comment by onox** [comment]
I have been rather busy with university and other things, so can't promise to work on it. You could start by searching for some sound that is heard when the engine is starting to get damaged. (Lowered RPM?)

What we could do is create some interpolation table in XML that tells how much points you get for a certain RPM.

Furthermore, I think we should lower the RPM a certain amount depending on how many points you have above a certain threshold. (another interpolation table giving a RPM reduction offset depending on collected points).

About the Aircraft Options dialog: I think it should be simplified a bit. We're starting to have a bit too many "Enable feature A", "Enable feature B", etc. checkbuttons. Personally I would combine "Save fuel level between sessions", "Save battery charge between sessions", and "Save positions of all switches between sessions". Combine them to "Save state of aircraft between sessions", because a user either want to resume its state when they restart FG or they don't. "Allow fuel contamination", "Allow oil management", and "Allow carburetor icing" basically say whether the user wants bad stuff to happen to the engine, so they could be combined to "Allow engines failures".

Just my 2 cents.

**Comment by onox** [comment]
Btw, I notice that I can no longer enable the tiedowns by clicking them in external view. I can still enable them via the Aircraft Options dialog.

**Comment by onox** [comment]
It seems @dg-505 accidentally removed `<not>` around `<property>/sim/current-view/internal</property>` when he simplified the expressions as I suggested (see 85a42173f0db40282f653ada481f16379ce4a107).

**Comment by onox** [comment]
I'm fixing it in the release branch.

**Comment by onox** [comment]
Fixed in https://github.com/Juanvvc/c172p-detailed/tree/release/2016.4

**Comment by onox** [comment]
Also cherry-picked to `master`.

**Comment by dg-505** [comment]
> It seems @dg-505 accidentally removed `<not>` around 


Oups, sorryy  😞

**Comment by gilbertohasnofb** [comment]
> I have been rather busy with university and other things, so can't promise to work on it. 

No problems at all

> You could start by searching for some sound that is heard when the engine is starting to get damaged. (Lowered RPM?)

:+1:


> What we could do is create some interpolation table in XML that tells how much points you get for a certain RPM.
> 
> Furthermore, I think we should lower the RPM a certain amount depending on how many points you have above a certain threshold. (another interpolation table giving a RPM reduction offset depending on collected points).

Sounds good.

> About the Aircraft Options dialog: I think it should be simplified a bit. 

Agreed. I will open a new issue for it and take a look at it soon. I propose the following:

- [ ] "save state of aircraft between sessions" (this would take care of fuel, battery, switches, etc.)
- [ ] "enable complex engine procedures and failures" (oil management, carb icing, engine damage, fuel contamination)

What od you think? "Enable complex engine procedures and failures" sounds better to me than "Allow engines failures", as failure has this connotation of something that happens randomly in a sim (e.g. gyro failure) and not something you caused it yourself. When I get this done, I would also document all these things in the wiki.

Still about the aircraft dialog, what about reducing it further:

- do we really need a toggle for the shadow? Can't it be enabled all the time? I don't see the point of this...
- "show registration in panel" is another thing I would remove

**Comment by dg-505** [comment]
> "show registration in panel" is another thing I would remove

IIRC the label with the registration on the panel shows your multiplayer callsign* which is in most times different to the ID of the livery. I can understand that some users want to see their mp callsign on the panel while others don't. Hence I see the purpose behind this option. 

*Correct me if I'm wrong here

**Comment by gilbertohasnofb** [comment]
> IIRC the label with the registration on the panel shows your multiplayer callsign* which is in most times different to the ID of the livery.

Yeah, I get that, but the thing is those liveries do have a "hard coded" registration and so does some of their panel textures (e.g. default livery 4X-CHV)). Having 2 registrations in the panel (one referring to the texture, one to the MP callsign) does look horrible IMO, but then again I have zero interest in MP.

As for the * in your message above, the panel sign actually shows the registration which you set via the aircraft menu, and I don't think it automatically shows your MP callsign (though maybe I am wrong here, maybe this is done automagically). Moreover, we have two liveries (custom callsign 1 and 2) which are made for that purpose, in which the texture lacks any registration both in the panel and outside, e.g.:

![fgfs-screen-002](https://cloud.githubusercontent.com/assets/5700795/20342583/59f7ce6a-abeb-11e6-89c1-978e12ff4794.png)

![fgfs-screen-003](https://cloud.githubusercontent.com/assets/5700795/20342584/59f8b870-abeb-11e6-881a-fa90e27b709f.png)

This is done automatically: when you select those liveries, the callsign will be displayed both in the cockpit and outside, and when you change for a regular livery those will be removed. The thing is we have this override toggle which allows the user to add the callsign to the panel of regular liveries as well.

But yeah, maybe we should just leave it as it is now, it's really not a big deal.

**Comment by onox** [comment]
"Enable complex engine procedures and failures" is good too :+1:

**Comment by gilbertohasnofb** [comment]
@onox what about the shadow, can I just remove that option as well?

**Comment by gilbertohasnofb** [comment]
Please see https://github.com/Juanvvc/c172p-detailed/issues/866

**Comment by onox** [comment]
I think we can remove that one as well.

**Comment by onox** [comment]
Would you prefer removing the "show registration in panel" checkbutton and remove it in the cockpit panel as well? This way it is never visible in the cockpit panel, but it is on the fuselage if the user chooses one of the custom callsign liveries.

**Comment by onox** [comment]
...or display it in the cockpit panel as well if the livery is one of the custom callsign liveries.

**Comment by onox** [comment]
(I have updated my previous question because it was a bit confusing)

**Comment by gilbertohasnofb** [comment]
@onox the current behaviour is: the registration shows in the fuselage only in if the user selects the liveries "Custom Callsign 1" or "Custom Callsign 2". When those are selected, the aircraft dialog option for showing callsign in the panel is automatically enabled, and when selecting any other livery it becomes disabled. The checkbox is used only in the case when the user is flying one of the non-custom callsign liveries but still wants to display it in the cockpit, thus overriding this automatic system. What I had proposed was to remove the checkbox thus removing the user's ability to override the system, since I find it useless and ugly, but then again I don't do MP and some people may find it useful. So I don't know, maybe we should just leave it there? The aircraft dialog is substantially smaller in the new PR I opened some short time ago

**Comment by onox** [comment]
> What I had proposed was to remove the checkbox thus removing the user's ability to override the system, since I find it useless and ugly, but then again I don't do MP

I do MP, but I find it useless and ugly too.

**Comment by onox** [comment]
Perhaps we could just remove the checkbutton but leave the ability to override in the code and see how many users complain on the forums.

**Comment by gilbertohasnofb** [comment]
> I do MP, but I find it useless and ugly too.

So in that case, do you think many people use this feature? I don't want to dictate how people should play with the toy we create, but I also don't like useless or ugly stuff.

> Perhaps we could just remove the checkbutton but leave the ability to override in the code and see how many users complain on the forums.

Let's do that, it would be easy to revert the commit if ever necessary. I will fix it in the already opened PR (there are two PR opened in case you'd have time to take a look at them)

**Comment by onox** [comment]
> So in that case, do you think many people use this feature? I don't want to dictate how people should play with the toy we create, but I also don't like useless or ugly stuff.

I don't know. It's kinda difficult to look in other's cockpits on MP :stuck_out_tongue_winking_eye: I see people rarely using the custom callsign liveries though. That of course doesn't tell us anything about whether a user has ticked that "show registration on panel" checkbutton.

We can decide to just remove the checkbutton and see if a Make The C172P Great Again movement arises on the forum.

**Comment by gilbertohasnofb** [comment]
> We can decide to just remove the checkbutton and see if a Make The C172P Great Again movement arises on the forum.

:laughing: sounds good! I also noticed that the callsigns are not persistent among sessions, so I am fixing that together. I will push the commits to https://github.com/Juanvvc/c172p-detailed/pull/869 very soon

**Comment by onox** [comment]
About the callsigns: do we want to sneak in some code that randomizes the callsign if it is "callsign"? New users with the callsign "callsign" will prevent other new users from using MP (because there can be only one user called "callsign").

I had already highlighted this problem on the mailing list, but there was no response from a core dev saying "I'm gonna fix this!".

This is of course very sneaky haha and just a workaround :laughing: , but core devs have no interest in MP :cry: 

Code is:

```diff
diff --git a/Nasal/c172p.nas b/Nasal/c172p.nas
index df72ae3..401d4df 100644
--- a/Nasal/c172p.nas
+++ b/Nasal/c172p.nas
@@ -561,6 +561,16 @@ var dialog_battery_reload = func {
 }
 
 setlistener("/sim/signals/fdm-initialized", func {
+    # Randomize callsign of new users to avoid them blocking
+    # other new users on multiplayer
+    if (getprop("/sim/multiplay/callsign") == "callsign") {
+        var digit = func {
+            return math.round(rand()*9);
+        };
+        var new_callsign = "FG-" ~ digit() ~ digit() ~ digit() ~ digit();
+        setprop("/sim/multiplay/callsign", new_callsign);
+    };
+
     # Use Nasal to make some properties persistent. <aircraft-data> does
     # not work reliably.
     aircraft.data.add("/sim/model/c172p/immat-on-panel");
```

**Comment by onox** [comment]
@gilbertohasnofb There's a message on the mailing list about splash screens called "Hangar thumbnails, screenshots, splash-screens":

> - redesign the splash screen to be more modern looking, with the -set.xml description and authors info automatically rendered on top (as we already do for name of program and startup info)

I prefer the splash screens to look clean (minus the http:// url at the top). Description and authors are already visible in the launcher.

**Comment by gilbertohasnofb** [comment]
>  do we want to sneak in some code that randomizes the callsign if it is "callsign"?

Yep, I think that's a good idea! But I just realized that we are using `/sim/model/immat` and not `/sim/multiplay/callsign`, which explains why my sim was showing a blank callsign instead of "callsign". How do you want to solve this? Should this function set both properties to the same value? And wouldn't you like to make this commit yourself to the PR I linked above, since this is solely your work? I don't mind doing it though

**Comment by onox** [comment]
The multiplayer system uses `/sim/multiplay/callsign` so that needs to be randomized. It could in addition set `/sim/model/immat` as well.

**Comment by onox** [comment]
Although it may be better to leave it blank (empty string) to save space in the MP packets.

**Comment by wkitty42** [comment]
On 11/16/2016 01:19 PM, onox wrote:
> I prefer the splash screens to look clean (minus the http:// url at the
> top). Description and authors are already visible in the launcher.


but they are not visible if users start via their own command line without any 
sort of launcher at all... automatically rendering them in the splash screen 
still allows for recognition if such is desired...

2 centavos

**Comment by gilbertohasnofb** [comment]
> I prefer the splash screens to look clean (minus the ugly yellow http:// url at the top). Description and authors are already visible in the launcher.

I am all for removing that ugly url, it has always bothered me.

> automatically rendering them in the splash screen still allows for recognition if such is desired...

I personally can live without that. But if you guys want, we can add an "About" dialog to the c172p menu with some info about the project, the team, the aircraft, links, etc..

**Comment by onox** [comment]
I would rather see the authors in an "About" dialog than on a splash screen.

**Comment by gilbertohasnofb** [comment]
> I would rather see the authors in an "About" dialog than on a splash screen.

:+1: I will open an issue for that and I can work on it later

**Comment by gilbertohasnofb** [comment]
See: https://github.com/Juanvvc/c172p-detailed/issues/870

**Comment by legoboyvdlp** [comment]
Confirm this is not implemented in the complex engine procedures?

**Comment by onox** [comment]
Engine damage has not been implemented so far. Do you want to work on it?

One way to implement it::

> What we could do is create some interpolation table in XML that tells how much points you get for a certain RPM.
>
> Furthermore, I think we should lower the RPM a certain amount depending on how many points you have above a certain threshold. (another interpolation table giving a RPM reduction offset depending on collected points).

**Comment by legoboyvdlp** [comment]
I've some ideas, yes -> but I'll be working on #751 first before I even think of this...

**Comment by legoboyvdlp** [comment]
Thinking out loud how you would do this ->

I'd probably simulate the actual damage in Nasal using a single setprop, then use an interpolation table to lower the RPM.

**Comment by legoboyvdlp** [comment]
Just a quick update here,
After an hours work, I've managed to make the engine kill itself if the oil pressure drops below 45psi, except when starting up, as well as coughing when it drops below 50psi. 

Also, I now have a damage simulation, where a nasal controller adds damage levels according to the following rule:

1 point of damage per 100RPM over limit added per second, so the damage increases faster the more you go over the limit

I'll need help with the interpolation table for the RPM offset, as I wouldn't be certain how much to drop the RPM as damage increases.

**Comment by onox** [comment]
> The main problem with over-RPM is probably stress on the rotating parts, correct?

That was the idea yes. 

> Should we prevent engine starting if the oil temperature is too low --> require you to heat the engine / sump with a heater? (optional, obviously!)

@gilbertohasnofb Does the c182 project has a heater that we can use?

> Also, should the engine damage be on the complex engine checkbox, or should all pilots be required to operate the engine properly to avoid damage? :P I take it the former...

Personally I don't like these extra checkboxes (just always simulate everything), but the majority has decided they want it.

**Comment by onox** [comment]
You can define the interpolation table in XML, maybe the easiest way.

**Comment by legoboyvdlp** [comment]
Hi Onox,
Thank you for your help 👍 😃 
So far I'm troubleshooting the oil pressure to make sure startup still works, once that is finished I'll troubleshoot the damage controller, then add the interpolation table :)

**Comment by onox** [comment]
You should also make sure things work in the flight recorder (replay). When the user replays the flight, the flight recorder writes the recorded properties and assumes that there's no Nasal that interferes. In the `master` branch I think the oil management code doesn't take that into account IIRC.

**Comment by legoboyvdlp** [comment]
Thanks - I'll do that :)

**Comment by legoboyvdlp** [comment]
@onox sorry to disturb you, just to let you know I've completed the Nasal / property rules part. I only need to create the interpolation table: for that, I would need some suggestions as to how damage levels (1 per 100RPM over the limit per second over the limit) affect RPM, and what level to kill the engine completely at.

**Comment by onox** [comment]
Does the POH contain any information about how regularly do an engine inspection?

**Comment by legoboyvdlp** [comment]
Checking.

**Comment by legoboyvdlp** [comment]
By the way, I see the C182 has a far more advanced oil management simulation - see engine.nas I wonder if @HHS81 would let us reuse his code for the C172P?

**Comment by legoboyvdlp** [comment]
@onox 
Before each flight of the day, the aircraft is checked for leaks and oil, fuel level, the air filter is checked, and the fuel drain is opened to remove sediment, water, etc

At 25 hours, new, rebuilt, or overhauled engines have their oil drained and renewed: aircraft that have "oil filter screens" must have this done every 25 hours no matter the age.

At 50 hours, the primer, throttle, mixture, spark plugs and ignition are checked, air filter is checked, and aircraft with "external full flow oil filters", i.e. no filter screens have their oil drained and renewed. The exhaust and cooling are also checked.

At 100 hours, the electrical connections are checked, the spark plugs are checked again and replaced if necessary, magnetos, cylinders, and engine injectors are checked. 

At 400 hours the cylinder valves are replaced.

That's all I found in the POH.

**Comment by onox** [comment]
@gilbertohasnofb implemented the oil management at the time. If the more advanced simulation can be added to the c172p that would be nice.

**Comment by HHS81** [comment]
@legoboyvdlp 
It is GNU GPL, so.... :smile:

**Comment by legoboyvdlp** [comment]
:)
Thanks, HHS!

**Comment by gilbertohasnofb** [comment]
> @gilbertohasnofb implemented the oil management at the time. If the more advanced simulation can be added to the c172p that would be nice.

As I wrote elsewhere, Lego, I did implement some sort of engine killing feature due to low oil but that's quite rudimentary. Basically the engine starts coughing below a certain threshold and dies below another lower threshold. If you implement engine damage, please remove the code that kills the engine (coughing should be fine though), it's all located in `engine.nas`. Let me know if you need a hand understanding what I did.

**Comment by legoboyvdlp** [comment]
I see exactly what you did, it is simple, but I like simple :) 

I've successfully managed to readapt your code for low pressure and high pressure. The engine will start to cough if you leave it at a very low idle speed - perhaps 400RPM for too long. It will die at about 35psi, unless the engine is in a starting state, which I had to hack to prevent constant coughing when starting...

I also implemented a rudimentary damage controller that adds one damage level per 100RPM over the redline per second. I only need information about how many levels it should take for the RPM to drop, and how many for the engine to die.

Finally, I'm right now working on adding the more advanced simulation of oil management from the C182S with Heiko's permission.

**Comment by legoboyvdlp** [comment]
@gilbertohasnofb I'd appreciate it if you could test the oil system, and coughing when the pressure is too low / high? 
That's the first part, the second part is to make the RPM drop as damage increases.

**Comment by gilbertohasnofb** [comment]
@legoboyvdlp I am not sure I understand you, have you pushed some modifications to your clone and you want me to test those or you want me to make some tests with the current master here? Also, I believe that with the current implementation pressure, temp and coughing are all consequences of low oil level and not directly connected to each other.

**Comment by legoboyvdlp** [comment]
Hello Gilberto, yes, I meant my fork. I've added oil management from the 182 and added coughing when the engine is low on oil pressure (not oil level, but pressure! I.E. the pump is not supplying enough due to low engine speed).

**Comment by gilbertohasnofb** [comment]
Ok, what branch in your repo?

**Comment by legoboyvdlp** [comment]
issue-124

**Comment by legoboyvdlp** [comment]
Essentially, I just would like a second reviewer to make sure I didn't do anything stupid and it works properly :)

**Comment by gilbertohasnofb** [comment]
@legoboyvdlp a couple of things:

- I think it might be easier if you open a PR and just put a warning such as :warning: WORK IN PROGRESS! or anything like that, so that people don't merge it by accident. This way we can use GitHub interface to see the diffs and comment the code
- I lowered the oil level to a low value of `1.99` and the engine coughing sounded really buggy
- I noticed a couple of properties that don't change at all. In the image below, the ones in red do not change and are stuck at 60 (compare that with the yellow temp property which does change), is that because of some of your modifications or was this already a previous bug? See:
 
![fg-1](https://user-images.githubusercontent.com/5700795/36614028-3ddb4b02-18d3-11e8-9693-4054698a15c0.jpg)

- since this involves a fair bit of coding and simulation, I think that @onox would be the best person to judge it. Besides that, I am already a bit overwhelmed by having too much on my hands at the moment both in RL as well as with the feedback on the 3D model and reworking a texture here and there and I am afraid I won't be able to help too much :cry: I am glad to help you understand any of the previous code I wrote for the oil consumption though if you need it (though I think you already got how that works)

**Comment by legoboyvdlp** [comment]
Thanks :)

I may have an idea what is causing the stuck properties, so I will attemp to fix them tomorrow.

As for the low oil level, at 2 quarts your engine likely would stop working... :) I will, however, reduce the coughing.
